### ISSUE TITLE
login-disabled role: log out users with this role on next safe request

### ISSUE THREAD

**Issue by Toutouwai** [processwire/processwire-requests#243]
### Short description of the enhancement

PW 3.0.86 introduced the login-disabled role: https://processwire.com/blog/posts/pw-3.0.86/
This feature works nicely for preventing users with that role from logging in, but it would be even better if it forced the logout of users with that role who are currently logged in. There are probably several use cases for this, but an example of when this would be useful is if you need to do some work on a live site and want to prevent any edits being made while the work is in progress.

It wouldn't be good to force the _immediate_ logout of all login-disabled users because this could result in lost data if the user is in the process of submitting a form. So this request is for a kind of "smart logout" feature that would force the logout of the user on the next request when it is safe to do so. Maybe the next request when there is no POST data in $input?

**Comment by Toutouwai** [comment]
An additional request relating to the login-disabled role: could the message shown to the user with this role who attempts to login be a bit more informative?

It's potentially confusing to the user to see the same "Login failed" message they see if they enter an invalid username or password, and may result in them attempting to reset the password and additional frustration when that still doesn't allow them to login. Could the message be "Login disabled" instead?

**Comment by teppokoivula** [comment]
> It's potentially confusing to the user to see the same "Login failed" message they see if they enter an invalid username or password, and may result in them attempting to reset the password and additional frustration when that still doesn't allow them to login. Could the message be "Login disabled" instead?

While I agree with the potential for confusion, it's worth noting that there are security implications here, though admittedly pretty small ones: if the system was to distinguish between non-existent accounts and accounts with logging in (currently) disabled, it would be possible to collect a list of usernames and use them in other sorts of attacks.

OWASP, for an example, [recommends as a best practice](https://www.owasp.org/index.php/Testing_for_User_Enumeration_and_Guessable_User_Account_(OWASP-AT-002)) that one should "verify that the application answers in the same manner for every client request that produces a failed authentication."

In my opinion it would be best to not make this change (by default). Rather I think that it would be best to add a custom hook or similar solution if and when this is deemed necessary.

---

For the record, I have no strong opinion either way on the first part of the request. It really depends on the interpretation of "login disabled". If we read it literally, it could (for an example) be used to create "single-session users": allow the user to log in initially, but then add this role, so that they can continue making changes for the current session but won't be able to log in again.

Valid use cases for that include editing your own posts, comments, or other kinds of submissions for a period of time – in this case defined by the length of the session.

**Comment by Toutouwai** [comment]
There's no security issue that I can see. The "Login disabled" message would only be shown in the case that the user would have otherwise successfully logged in had they not been given the login-disabled role (i.e. the entered username and password are correct).

**Comment by szabeszg** [comment]
> if the system was to distinguish between non-existent accounts and accounts with logging in (currently) disabled, it would be possible to collect a list of usernames and use them in other sorts of attacks.

In the case of a temporary state of the maintenance use case described above, the chance of this is pretty low I guess, especially in the case of low traffic sites.

If this request is all about putting the site into maintenance state then maybe ProcessWire should have support for it out of the box, and not just via hooks and custom coding but via the admin.

**Comment by teppokoivula** [comment]
> There's no security issue that I can see. The "Login disabled" message would only be shown in the case that the user would have otherwise successfully logged in had they not been given the login-disabled role (i.e. the entered username and password are correct).

Agreed – didn't consider this point – so yeah, forget my earlier comment re security. Not an issue :)

What @szabeszg said about maintenance mode is what I was going to suggest as well, but then again there are already modules for that, so whether this really needs to be a core feature is, in my opinion, questionable. I guess the question should be whether most of ProcessWire users would benefit from the ability to put the site in maintenance mode?

**Comment by Toutouwai** [comment]
I see this as being distinct from a maintenance mode. Maintenance mode is about excluding general visitors from the front-end while maintenance is being done. The login-disabled role (and this request) is about excluding PW users from the back-end for a range of different reasons.

I took a look at look at how difficult it would be to automatically log out users in a way that doesn't result in losing data they might be in the process of submitting. Seems it's quite easy - I hooked after `Page::render()` and if the user should have their login disabled I do a `$session->logout()` then a `$session->redirect()` back to the current URL (so they can potentially continue where they left off if/when their login is re-enabled). If the user is being logged out from a back-end URL I do a `$session->warning()` message explaining what has happened.

Quite easy to put into a module for a range of different purposes (logout specific users, logout specific roles, logout all non-superusers, etc), so not essential for this to be in the core. For the core, to me it boils down to what is most useful to most users, and I think generally speaking if you want to disable the login for a user then you don't want that user logged in full stop. But this could be configurable perhaps.

**Comment by teppokoivula** [comment]
> But this could be configurable perhaps.

In this case it would be a potentially breaking change, so if this behaviour is changed, this might be a good idea.

For the record, in the LoginScheduler module [I've hooked to Session::isValidSession()](https://github.com/teppokoivula/LoginScheduler/blob/master/LoginScheduler.module#L55), and it's working fine for the scope of that module. Sure, it's possible for a POST request (save, edit) to fail, but that's literally the intention there. In other words it works pretty much exactly the way you've described here, with the difference that logout is instantaneous (i.e. submitted data can be lost) :)
